### PR TITLE
fix: removed sso client registration/deregistration

### PR DESCRIPTION
**Pull request by ziccardi**
## Checklist (Definition of Done)

- [ ] All acceptance criteria specified in JIRA have been completed
- [ ] Unit and integration tests added that prove the fix is effective or the feature works (tested against emulated and non-emulated OCM environment)
- [ ] Documentation added for the feature
- [ ] CI and all relevant tests are passing
- [ ] Code Review completed
- [ ] Verified independently by reviewer
- [ ] Required metrics/dashboards/alerts have been added (or PR created).
- [ ] Required Standard Operating Procedure (SOP) is added.
- [ ] JIRA has created for changes required on the client side

### CODE REVIEW

**Comment by machi1990** [comment]
> It will also require removing this migration `https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/blob/main/internal/kafka/internal/migrations/20210322131730_add_kafka_sso_clientid_secret.go`

I'll not say removing the migration as that won't have any effect. What we need is another migration that will remove the columns.

I think what will need to be done is:
In the first PR:
- (a)first, stop creating the kafka client and remove the register Kafka client method everywhere
- (b) add a reconciler (background worker) that deregister the Kafka client id from mas-sso for the Kafkas that have them (i.e existing Kafkas) 

In another followup PR:
- (c) remove the deregister kafka client method everywhere
- (d) add a migration to remove the kafka client id / client secret from the database
- (e) remove client id and client secret from the KafkaRequest

That's is if we want to be idealist and let the code handle everything. 

Optionally, we can everything in one PR and do part (c) manually in Stage and Prod.
Then if we do that, the things to do will look like:
- (a)first, stop creating the kafka client and remove the register and deregister Kafka client method )everywhere
- (b) add a migration to remove the kafka client id / client secret from the database
- (c) remove client id and client secret from the KafkaRequest
- (d) ~~add a reconciler (background worker) that deregister the Kafka client id from mas-sso for the Kafkas that have them (i.e existing Kafkas)~~ This will have to be done manually

**Comment by akoserwal** [comment]
> > It will also require removing this migration `https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/blob/main/internal/kafka/internal/migrations/20210322131730_add_kafka_sso_clientid_secret.go`
> 
> I'll not say removing the migration as that won't have any effect. What we need is another migration that will remove the columns.
> 
> I think what will need to be done is: In the first PR:
> 
> * (a)first, stop creating the kafka client and remove the register Kafka client method everywhere
> * (b) add a reconciler (background worker) that deregister the Kafka client id from mas-sso for the Kafkas that have them (i.e existing Kafkas)
> 
> In another followup PR:
> 
> * (c) remove the deregister kafka client method everywhere
> * (d) add a migration to remove the kafka client id / client secret from the database
> * (e) remove client id and client secret from the KafkaRequest
> 
> That's is if we want to be idealist and let the code handle everything.
> 
> Optionally, we can do all the required in one PR and do part (c) manually in Stage and Prod. Then if we do that, the things to do will look like:
> 
> * (a) stop creating the kafka client and remove the register and deregister Kafka client method )everywhere
> * (b) add a migration to remove the kafka client id / client secret from the database
> * (c) remove client id and client secret from the KafkaRequest
> * (d) ~add a reconciler (background worker) that deregister the Kafka client id from mas-sso for the Kafkas that have them (i.e existing Kafkas)~ This will have to be done manually

(a)first, stop creating the kafka client and remove the register Kafka client method everywhere
(b) add a migration to remove the kafka client id / client secret from the database

Clients will be removed manually.

**Comment by machi1990** [comment]
> (a)first, stop creating the kafka client and remove the register Kafka client method everywhere 
(b) add a migration to remove the kafka client id / client secret from the database
> 
> Clients will be removed manually.

Awesome. We also have to remove the 
```
SsoClientID                      string `json:"sso_client_id"`
SsoClientSecret                  string `json:"sso_client_secret"`
```	
from the `kafka_request_types.go` file.
which is part (c) 
> (c) remove client id and client secret from the KafkaRequest

**Comment by ziccardi** [comment]
> > (a)first, stop creating the kafka client and remove the register Kafka client method everywhere
> > (b) add a migration to remove the kafka client id / client secret from the database
> > Clients will be removed manually.
> 
> Awesome. We also have to remove the
> 
> ```
> SsoClientID                      string `json:"sso_client_id"`
> SsoClientSecret                  string `json:"sso_client_secret"`
> ```
> 
> from the `kafka_request_types.go` file. which is part (c)
> 
> > (c) remove client id and client secret from the KafkaRequest

You mean from `kafka_request.go`? If yes, they have been already removed in this PR

File: internal/kafka/internal/services/kafka.go [hunk 1]
```diff
@@ -1089,8 +1076,6 @@ func buildManagedKafkaCR(kafkaRequest *dbapi.KafkaRequest, kafkaConfig *config.K
 
 	if keycloakConfig.EnableAuthenticationOnKafka {
 		managedKafkaCR.Spec.OAuth = managedkafka.OAuthSpec{
-			ClientId:               kafkaRequest.SsoClientID,
-			ClientSecret:           kafkaRequest.SsoClientSecret,
```
Comment (machi1990) [inline internal/kafka/internal/services/kafka.go:1093]: These https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/blob/75fae9d3d049e58c06a7f70858de4fad03cf4a5c/openapi/kas-fleet-manager-private.yaml#L287-L290 needs to be removed as well. 
What do you think? @akoserwal @ziccardi

File: openapi/kas-fleet-manager-private-admin.yaml [hunk 2]
```diff
@@ -255,8 +255,6 @@ components:
               type: string
             subscription_id:
               type: string
-            sso_client_id:
-              type: string
```
Comment (machi1990) [inline openapi/kas-fleet-manager-private-admin.yaml:259]: I also think these https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/blob/0acbe9de6cbf46d37b729251b736667a602bf7e7/openapi/kas-fleet-manager-private.yaml#L287-L290 can be removed

Comment (akoserwal) [reply openapi/kas-fleet-manager-private-admin.yaml:259]: yes

File: internal/kafka/internal/services/kafka.go (hunk 1 above)
Comment (akoserwal) [reply internal/kafka/internal/services/kafka.go:1093]: yes

File: openapi/kas-fleet-manager-private.yaml [hunk 3]
```diff
@@ -275,19 +275,15 @@ components:
...
                   $ref: "#/components/schemas/ManagedKafkaCapacity"
                 oauth:
                   type: object
                   required:
                     - maximumSessionLifetime
                   properties:
-                    clientId:
-                      type: string
-                    clientSecret:
-                      type: string
```
Comment (machi1990) [inline openapi/kas-fleet-manager-private.yaml:290]: Let's bump the OpenAPI version https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/blob/6e30f6b93614bfd64a4c1cb05f0a8298ba13bd70/openapi/kas-fleet-manager-private.yaml#L8 following this change

File: openapi/kas-fleet-manager-private-admin.yaml [hunk 4]
```diff
@@ -255,8 +255,6 @@ components:
               type: string
             subscription_id:
               type: string
-            sso_client_id:
```
Comment (machi1990) [inline openapi/kas-fleet-manager-private-admin.yaml:258]: let's bump the version following this change https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/blob/6e30f6b93614bfd64a4c1cb05f0a8298ba13bd70/openapi/kas-fleet-manager-private-admin.yaml#L5

File: openapi/kas-fleet-manager-private.yaml [hunk 5]
```diff
@@ -275,19 +275,15 @@ components:
...
                 capacity:
                   $ref: "#/components/schemas/ManagedKafkaCapacity"
                 oauth:
                   type: object
                   required:
                     - maximumSessionLifetime
                   properties:
-                    clientId:
-                      type: string
-                    clientSecret:
```
Comment (machi1990) [inline openapi/kas-fleet-manager-private.yaml:289]: @MikeEdgar @shawkins from my understanding, FSO is not using these fields. This PR is removing the field from the database and hence from the ManagedKafka#Spec#Oauth field. Can they safely be removed? Or the fields are current being used in FSO?

Comment (MikeEdgar) [reply openapi/kas-fleet-manager-private.yaml:289]: @machi1990 they've been made optional, but the code is still referencing them and they are present in the API spec. It's safe to no longer send the properties. We can completely remove them in a future FSO change once they're no longer sent.

https://github.com/bf2fc6cc711aee1a0c2a/kas-fleetshard/pull/700

Comment (machi1990) [reply openapi/kas-fleet-manager-private.yaml:289]: Thanks for confirming

File: internal/kafka/internal/workers/clusters_mgr.go [hunk 6]
```diff
@@ -409,12 +409,6 @@ func (c *ClusterManager) reconcileDeprovisioningCluster(cluster *api.Cluster) er
 }
 
 func (c *ClusterManager) reconcileCleanupCluster(cluster api.Cluster) error {
-	glog.Infof("Removing Dataplane cluster %s IDP client", cluster.ClusterID)
-	keycloakDeregistrationErr := c.OsdIdpKeycloakService.DeRegisterClientInSSO(cluster.ID)
```
Comment (akoserwal) [inline internal/kafka/internal/workers/clusters_mgr.go:413]: This will be required

File: internal/kafka/internal/workers/clusters_mgr.go [hunk 7]
```diff
@@ -409,12 +409,6 @@ func (c *ClusterManager) reconcileDeprovisioningCluster(cluster *api.Cluster) er
...
-	if keycloakDeregistrationErr != nil {
```
Comment (akoserwal) [inline internal/kafka/internal/workers/clusters_mgr.go:414]: This is required

File: internal/kafka/internal/workers/clusters_mgr_test.go [hunk 8]
```diff
@@ -2884,11 +2866,7 @@ func TestClusterManager_reconcileCleanupCluster(t *testing.T) {
 						return nil
 					},
 				},
-				osdIDPKeycloakService: &sso.KeycloakServiceMock{
-					DeRegisterClientInSSOFunc: func(kafkaNamespace string) *apiErrors.ServiceError {
```
Comment (akoserwal) [inline internal/kafka/internal/workers/clusters_mgr_test.go:2888]: This is required

File: internal/kafka/internal/workers/clusters_mgr.go (hunk 6 above)
Comment (machi1990) [reply internal/kafka/internal/workers/clusters_mgr.go:413]: +1

File: pkg/services/sso/keycloak_service_builder.go [hunk 9]
```diff
@@ -8,34 +8,103 @@ import (
...
 
+type OSDKeycloakServiceBuilder interface {
+	WithRealmConfig(realmConfig *keycloak.KeycloakRealmConfig) OSDKeycloakServiceBuilder
+	Build() OSDKeycloakService
+}
+
+type keycloakServiceBuilderSelector struct {
+}
+
+func (s *keycloakServiceBuilderSelector) ForOSD() OSDKeycloakServiceBuilderConfigurator {
```
[the remark this answers is not in the document]
Comment (akoserwal) [reply pkg/services/sso/keycloak_service_builder.go:40]: okay make sense. can you make the `OSDKeycloakInterface` to always use the `OSDClusterIDPRealm` (rhoas-kafka-sre)

File: pkg/services/sso/keycloak_service_builder.go [hunk 10]
```diff
@@ -8,34 +8,103 @@ import (
...
+	return &osdBuilderConfigurator{}
+}
+
+func (s *keycloakServiceBuilderSelector) ForKFM() KFMKeycloakServiceBuilderConfigurator {
+	return &keycloakBuilderConfigurator{}
+}
```
Comment (machi1990) [inline pkg/services/sso/keycloak_service_builder.go:46]: Why the `.ForKFM()` name? How is this method meant to be used by other fleet managers? KFM and connectors as well? 
This file is in the core service and with the name `.ForKFM()` it seems like it is specific for kas-fleet-manager, is this true?

@ziccardi @akoserwal

File: pkg/services/sso/keycloak_service_builder.go [hunk 11]
```diff
@@ -8,34 +8,103 @@ import (
...
+	builder.realmConfig = realmConfig
+	return builder
+}
+
+// Build returns an instance of KeycloakService ready to be used.
+// If a custom realm is configured (WithRealmConfig called), then always Keycloak provider is used
+// irrespective of the `builder.config.SelectSSOProvider` value
+func (builder *osdKeycloackServiceBuilder) Build() OSDKeycloakService {
+	return build(builder.config.SelectSSOProvider, builder.config, builder.realmConfig).(OSDKeycloakService)
+}
```
Comment (machi1990) [inline pkg/services/sso/keycloak_service_builder.go:87]: The comment is not true for a keycloak service used for the configuration of the identity provider. 
For this service, we should always use the rhoas-kafka-sre role irrespective of the provider type i.e there is no need to pass `builder.config.SelectSSOProvider` or have a `WithRealmConfig(...)` that takes a realm in argument.

File: pkg/services/sso/keycloak_service_builder.go [hunk 12]
```diff
@@ -8,34 +8,103 @@ import (
...
+
+func (builder *osdKeycloackServiceBuilder) WithRealmConfig(realmConfig *keycloak.KeycloakRealmConfig) OSDKeycloakServiceBuilder {
+	builder.realmConfig = realmConfig
+	return builder
+}
+
+func build(providerName string, keycloakConfig *keycloak.KeycloakConfig, realmConfig *keycloak.KeycloakRealmConfig) KeycloakService {
```
Comment (machi1990) [inline pkg/services/sso/keycloak_service_builder.go:94]: `providerName` is already in `keycloakConfig *keycloak.KeycloakConfig`, we can reduce the number of arguments by removing it and infer it from `keycoakConfig` whenever we reference its value

**Pull request merged**